source/Mlos.Model.Services.Client/OptimizerServiceEncoderDecoder.cs

```diff
@@ -0,0 +1,419 @@
+// -----------------------------------------------------------------------
+// <copyright file="OptimizerServiceEncoderDecoder.cs" company="Microsoft Corporation">
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root
+// for license information.
+// </copyright>
+// -----------------------------------------------------------------------
+
+using Mlos.Model.Services.Spaces;
+
+namespace Mlos.Model.Services.Client
+{
+    public class OptimizerServiceEncoder
+    {
+        public OptimizerService.SimpleHypergrid EncodeHypergrid(Hypergrid hypergrid)
+        {
+            OptimizerService.SimpleHypergrid instance = new OptimizerService.SimpleHypergrid();
+
+            // TODO: I have no idea if this is the right way to handle subgrids.
+            foreach (var subgridSet in hypergrid.Subgrids.Values)
+            {
+                foreach (var subgrid in subgridSet)
+                {
+                    instance.GuestSubgrids.Add(EncodeSubgridJoin(subgrid));
+                }
+            }
+
+            foreach (var dimension in hypergrid.Dimensions)
+            {
+                instance.Dimensions.Add(EncodeDimension(dimension));
+            }
+
+            instance.Name = hypergrid.Name;
+
+            return instance;
+        }
+
+        public OptimizerService.OptimizationProblem EncodeOptimizationProblem(OptimizationProblem problem)
+        {
+            OptimizerService.OptimizationProblem instance = new OptimizerService.OptimizationProblem();
+            instance.ParameterSpace = EncodeHypergrid(problem.ParameterSpace);
+            instance.ObjectiveSpace = EncodeHypergrid(problem.ObjectiveSpace);
+            foreach (var objective in problem.Objectives)
+            {
+                instance.Objectives.Add(EncodeObjective(objective));
+            }
+
+            if (problem.ContextSpace == null)
+            {
+                instance.EmptyContext = new OptimizerService.Empty();
+            }
+            else
+            {
+                instance.ContextSpace = EncodeHypergrid(problem.ContextSpace);
+            }
+
+            return instance;
+        }
+
+        public OptimizerService.Objective EncodeObjective(OptimizationObjective objective)
+        {
+            return new OptimizerService.Objective
+            {
+                Minimize = objective.Minimize,
+                Name = objective.Name,
+            };
+        }
+
+        public OptimizerService.GuestSubgrid EncodeSubgridJoin(SubgridJoin subgridJoin)

```

Comment:
What is a SubgridJoin?

Reply:
Found it, throughout the rest of the code it's called a JoinedSubgrid or a GuestSubgrid. It may be easier to map this to Python concepts if we keep the names the same.

Reply:
Re: Previous comment. I just realized that I actually encoded the join rather than the subgrid. Fixing now.

Reply:
[zaporter/grpc de5d987] SubgridJoin => JoinedSubgrid, Deleted most references to JSON serialization